File: backend/app/schemas/report.py
from pydantic import BaseModel
from typing import Optional, List, Any
from uuid import UUID
from datetime import datetime
from decimal import Decimal


class ReportCreate(BaseModel):
    device_hash: str
    incident_type_id: int
    description: Optional[str] = None
    latitude: Decimal
    longitude: Decimal
    gps_accuracy: Optional[Decimal] = None
    motion_level: Optional[str] = None  # low / medium / high
    movement_speed: Optional[Decimal] = None
    was_stationary: Optional[bool] = None
    village_location_id: Optional[int] = None


class ReportUpdate(BaseModel):
    rule_status: Optional[str] = None     # passed / flagged / rejected
    is_flagged: Optional[bool] = None


class ReportResponse(BaseModel):
    report_id: UUID
    device_id: UUID
    incident_type_id: int
    description: Optional[str]
    latitude: Decimal
    longitude: Decimal
    gps_accuracy: Optional[Decimal]
    motion_level: Optional[str]
    movement_speed: Optional[Decimal]
    was_stationary: Optional[bool]
    village_location_id: Optional[int]
    reported_at: datetime
    rule_status: str
    is_flagged: bool
    ai_ready: bool

    class Config:
        from_attributes = True


# ── Nested response schemas (imported inline to avoid circular deps) ──

class _EvidenceFileBrief(BaseModel):
    evidence_id: UUID
    file_url: str
    file_type: str
    media_latitude: Optional[Decimal] = None
    media_longitude: Optional[Decimal] = None
    is_live_capture: bool = False
    uploaded_at: datetime
    blur_score: Optional[Decimal] = None
    tamper_score: Optional[Decimal] = None
    ai_quality_label: Optional[str] = None
    class Config:
        from_attributes = True


class _MLPredictionBrief(BaseModel):
    prediction_id: UUID
    trust_score: Optional[Decimal] = None
    prediction_label: Optional[str] = None
    confidence: Optional[Decimal] = None
    explanation: Optional[Any] = None
    model_type: Optional[str] = None
    is_final: bool = False
    evaluated_at: datetime
    class Config:
        from_attributes = True


class _PoliceReviewBrief(BaseModel):
    review_id: UUID
    police_user_id: int
    decision: str
    review_note: Optional[str] = None
    ground_truth_label: Optional[str] = None
    confidence_level: Optional[Decimal] = None
    reviewed_at: datetime
    class Config:
        from_attributes = True


class _AssignmentBrief(BaseModel):
    assignment_id: UUID
    police_user_id: int
    status: str
    priority: str
    assigned_at: datetime
    completed_at: Optional[datetime] = None
    class Config:
        from_attributes = True


class _IncidentTypeBrief(BaseModel):
    incident_type_id: int
    type_name: str
    severity_weight: Decimal
    class Config:
        from_attributes = True


class _DeviceBrief(BaseModel):
    device_id: UUID
    device_hash: str
    device_trust_score: Decimal
    total_reports: int
    trusted_reports: int
    flagged_reports: int
    class Config:
        from_attributes = True


class ReportDetailResponse(ReportResponse):
    evidence_files: List[_EvidenceFileBrief] = []
    ml_predictions: List[_MLPredictionBrief] = []
    police_reviews: List[_PoliceReviewBrief] = []
    assignments: List[_AssignmentBrief] = []
    incident_type: Optional[_IncidentTypeBrief] = None
    device: Optional[_DeviceBrief] = None


class ReportListResponse(BaseModel):
    reports: List[ReportResponse]
    total: int
    page: int
    per_page: int

File: mobile/lib/models/report.dart
/// Mirrors the backend `reports` table (submission payload).
class Report {
  final String? reportId;
  final String deviceHash;
  final String incidentTypeId;
  final String description;
  final double latitude;
  final double longitude;
  final double? gpsAccuracy;
  final String? motionLevel;    // low | medium | high
  final double? movementSpeed;
  final bool? wasStationary;
  final String? villageLocationId;
  final DateTime? reportedAt;
  final String? ruleStatus;
  final bool? isFlagged;

  const Report({
    this.reportId,
    required this.deviceHash,
    required this.incidentTypeId,
    required this.description,
    required this.latitude,
    required this.longitude,
    this.gpsAccuracy,
    this.motionLevel,
    this.movementSpeed,
    this.wasStationary,
    this.villageLocationId,
    this.reportedAt,
    this.ruleStatus,
    this.isFlagged,
  });

  factory Report.fromJson(Map<String, dynamic> json) {
    return Report(
      reportId: json['report_id'] as String?,
      deviceHash: json['device_hash'] as String? ?? '',
      incidentTypeId: json['incident_type_id'] as String,
      description: json['description'] as String,
      latitude: (json['latitude'] as num).toDouble(),
      longitude: (json['longitude'] as num).toDouble(),
      gpsAccuracy: (json['gps_accuracy'] as num?)?.toDouble(),
      motionLevel: json['motion_level'] as String?,
      movementSpeed: (json['movement_speed'] as num?)?.toDouble(),
      wasStationary: json['was_stationary'] as bool?,
      villageLocationId: json['village_location_id'] as String?,
      reportedAt: json['reported_at'] != null
          ? DateTime.tryParse(json['reported_at'] as String)
          : null,
      ruleStatus: json['rule_status'] as String?,
      isFlagged: json['is_flagged'] as bool?,
    );
  }

  /// Payload sent to POST /api/v1/reports
  Map<String, dynamic> toSubmitJson() => {
        'device_hash': deviceHash,
        'incident_type_id': incidentTypeId,
        'description': description,
        'latitude': latitude,
        'longitude': longitude,
        if (gpsAccuracy != null) 'gps_accuracy': gpsAccuracy,
        if (motionLevel != null) 'motion_level': motionLevel,
        if (movementSpeed != null) 'movement_speed': movementSpeed,
        if (wasStationary != null) 'was_stationary': wasStationary,
        if (villageLocationId != null) 'village_location_id': villageLocationId,
      };
}
